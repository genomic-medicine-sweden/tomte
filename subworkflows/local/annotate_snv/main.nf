//
// Annotating SNVs
//

include { ENSEMBLVEP_VEP       } from '../../../modules/nf-core/ensemblvep/vep'
include { RENAME_FILES         } from '../../../modules/local/rename_files'
include { TABIX_BGZIPTABIX     } from '../../../modules/nf-core/tabix/bgziptabix'
include { TABIX_TABIX          } from '../../../modules/nf-core/tabix/tabix'
include { GAWK                 } from '../../../modules/nf-core/gawk'
include { ENSEMBLVEP_FILTERVEP } from '../../../modules/nf-core/ensemblvep/filtervep'



workflow ANNOTATE_SNV {
    take:
    vcf                            //   channel:   [mandatory] [ val(meta), path(vcf) ]
    val_vep_genome                 // parameter: [mandatory] 'GRCh37' or 'GRCh38'
    val_vep_cache_version          // parameter: [mandatory] default: 110
    ch_vep_cache                   //   channel:   [mandatory] [ path(cache) ]
    ch_fasta                       //   channel:   [mandatory] [ val(meta), path(fasta) ]
    ch_vep_extra_files             //   channel:   [mandatory] [ path(files) ]
    ch_gene_panel_clinical_filter  //   channel:   [optional]  [ path(file) ]
    skip_clinical_filter           // parameter: [mandatory] default: true

    main:
        ch_versions = Channel.empty()

        // Annotate with VEP
        ENSEMBLVEP_VEP(
            vcf.map{ meta, vcf -> [ meta, vcf, [] ] },
            val_vep_genome,
            "homo_sapiens",
            val_vep_cache_version,
            ch_vep_cache,
            ch_fasta,
            ch_vep_extra_files
        )

        ENSEMBLVEP_VEP.out.vcf
            .multiMap { meta, vcf ->
                clinical: [ meta + [ set: "clinical" ], vcf ]
                research: [ meta + [ set: "research" ], vcf ]
            }
            .set { ch_clin_research_vcf }

        RENAME_FILES( ch_clin_research_vcf.research )

        TABIX_TABIX( RENAME_FILES.out.output )
        ch_vcf_research = RENAME_FILES.out.output.join(TABIX_TABIX.out.tbi)

        // Generate Clinical filter
        GAWK( ch_gene_panel_clinical_filter.map{it -> [[id:'hgnc'], it]}.collect(),
            [] )

        ch_hgnc_ids = GAWK.out.output.map{ meta, hgnc_ids -> [ hgnc_ids ] }

        //Filter results
        if ( !skip_clinical_filter ) {
            ENSEMBLVEP_FILTERVEP(
                ch_clin_research_vcf.clinical,
                ch_hgnc_ids
            )
            .output
            .set { ch_filtervep_out }

            TABIX_BGZIPTABIX( ch_filtervep_out )
            ch_vcf_clin = TABIX_BGZIPTABIX.out.gz_tbi.mix.collect()
            ch_versions = ch_versions.mix( ENSEMBLVEP_FILTERVEP.out.versions )
            ch_versions = ch_versions.mix( TABIX_BGZIPTABIX.out.versions )
        } else {
            ch_vcf_clin = Channel.empty()
        }

        ch_versions = ch_versions.mix( ENSEMBLVEP_VEP.out.versions.first() )
        ch_versions = ch_versions.mix( RENAME_FILES.out.versions )
        ch_versions = ch_versions.mix( GAWK.out.versions )
        ch_versions = ch_versions.mix( TABIX_TABIX.out.versions )

    emit:
        tab_gz          = ENSEMBLVEP_VEP.out.tab    // channel: [ val(meta), path(tab.gz) ]
        json_gz         = ENSEMBLVEP_VEP.out.json   // channel: [ val(meta), path(json.gz) ]
        report          = ENSEMBLVEP_VEP.out.report // channel: [ path(html) ]
        ch_vcf_clin     = ch_vcf_clin               // channel: [ val(meta), path(vcf.gz) path(tbi)]
        ch_vcf_research = ch_vcf_research           // channel: [ val(meta), path(vcf.gz) path(tbi)]
        versions        = ch_versions               // channel: [ path(versions.yml) ]
}
