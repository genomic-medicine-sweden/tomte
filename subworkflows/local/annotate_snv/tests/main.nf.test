nextflow_workflow {

    name "Test Workflow ANNOTATE_SNV"
    script "../main.nf"
    workflow "ANNOTATE_SNV"
    config "./nextflow.config"
    tag "subworkflow"
    tag "annotate_snv"

    setup {
        run("GUNZIP", alias: "GUNZIP_FASTA") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'fasta' ],
                    file("${projectDir}/test_data/grch37_chr21.fa.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("UNTAR", alias: "UNTAR_VEP_CACHE") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'vep_cache' ],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/vep_cache_and_plugins.tar.gz", checkIfExists: true)
                ])
                """
            }
        }
    }

    test("Should run without failures") {

        when {
            params {
                skip_vep = false
            }
            workflow {
                """
                input[0] = Channel.of([
                        [ probands:[ 'ACC5963A1' ], id:'finequagga' ],
                        file("${projectDir}/test_data/ACC5963A1_split_rmdup_info.vcf.gz", checkIfExists: true)
                        ])
                input[1] = 'GRCh37'
                input[2] = 107
                input[3] = UNTAR_VEP_CACHE.out.untar.map{meta, vep_cache -> vep_cache}.collect()
                input[4] = GUNZIP_FASTA.out.gunzip.collect()
                input[5] = channel.fromPath("https://raw.githubusercontent.com/nf-core/test-datasets/raredisease/reference/vep_files.csv", checkIfExists: true)
                input[6] = channel.fromPath("${projectDir}/test_data/drop_data/mock_gene_panel.tsv", checkIfExists: true)
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.tab_gz,
                    workflow.out.json_gz,
                    file(workflow.out.report.get(0)).name,
                    workflow.out.ch_vcf_clin.collect{path(it[1]).vcf.summary},
                    workflow.out.ch_vcf_research.collect{path(it[1]).vcf.summary},
                    workflow.out.versions
                ).match() }
            )
        }

    }

}
