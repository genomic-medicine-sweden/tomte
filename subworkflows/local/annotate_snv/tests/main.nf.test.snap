{
    "Should run without failures": {
        "content": [
            [
                
            ],
            [
                
            ],
            "ACC5963A1_split_rmdup_info_vep.vcf.gz_summary.html",
            [
                "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=113, phased=false, phasedAutodetect=false]"
            ],
            [
                "VcfFile [chromosomes=[chr21], sampleCount=1, variantCount=805, phased=false, phasedAutodetect=false]"
            ],
            [
                "versions.yml:md5,4ab5aab4abb17eb59948e84bc3caacab",
                "versions.yml:md5,751a6d20ce7cac841bd18b19e74189eb",
                "versions.yml:md5,767cd878bb512f92163193378381d5c9",
                "versions.yml:md5,8a206ac32b9dc4804f742ea5192e0f7e",
                "versions.yml:md5,9e92eb16a95a907da89ecabc16ce9247",
                "versions.yml:md5,c503a228cc6190d09a7fff5d07a59040"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-11T10:50:25.843804"
    }
}